src/Visitor/NameResolver.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\SemanticVersionChecker\Visitor;
+
+use PhpParser\Node;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\NodeVisitor\NameResolver as ParserNameResolver;
+use PhpParser\BuilderHelpers;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocNode;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode;
+use PHPStan\PhpDocParser\Ast\Type\TypeNode;
+use PHPStan\PhpDocParser\Ast\Type\UnionTypeNode;
+use PHPStan\PhpDocParser\Lexer\Lexer;
+use PHPStan\PhpDocParser\Parser\ConstExprParser;
+use PHPStan\PhpDocParser\Parser\PhpDocParser;
+use PHPStan\PhpDocParser\Parser\TokenIterator;
+use PHPStan\PhpDocParser\Parser\TypeParser;
+
+/**
+ * Extended Name Resolver that parse and resolve also docblock hintings
+ */
+class NameResolver extends ParserNameResolver
+{
+    /**
+     * @inheritDoc
+     */
+    public function enterNode(Node $node)
+    {
+        $return = parent::enterNode($node);
+
+        if ($node instanceof ClassMethod) {
+            $this->resolveDocBlockParamTypes($node);
+        }
+
+        return $return;
+    }
+
+    /**
+     * @param ClassMethod $node
+     * @return void
+     */
+    private function resolveDocBlockParamTypes(ClassMethod $node)
+    {
+        /** @var PhpDocNode $docNode */
+        $docNode = $this->getParsedDocNode($node);
+        if ($docNode) {
+            $result = [];
+            /** @var PhpDocTagNode[] $paramTags */
+            $paramTags = $docNode->getTagsByName('@param');
+            /** @var PhpDocTagNode $paramTag */
+            foreach ($paramTags as $paramTag) {
+                $paramNode = [
+                    'name' => $paramTag->value->parameterName ?? '',
+                    'type' => $this->parseType($paramTag->value->type),
+                ];
+                $result['params'][] = $paramNode;
+            }
+
+            /** @var PhpDocTagNode[] $returnTags */
+            $returnTags = $docNode->getTagsByName('@return');
+            /** @var PhpDocTagNode $returnTag */
+            $returnTag = array_shift($returnTags);
+            if ($returnTag) {
+                $result['return'] = $this->parseType($returnTag->value->type);
+            }
+            $node->setAttribute('docCommentParsed', $result);
+        }
+    }
+
+    /**
+     * Parse param or return type into array of resolved types
+     *
+     * @param TypeNode $type
+     * @return array
+     */
+    private function parseType($type)
+    {
+        $result = [];
+        if ($type instanceof UnionTypeNode) {
+            foreach ($type->types as $typeNode) {
+                $normalizedType = BuilderHelpers::normalizeType((string)$typeNode);
+                $resolvedType = $this->resolveType($normalizedType);
+                $result[] = $resolvedType;
+            }
+        } else {
```

Comment:
Note: Please ensure that `PHPStan\PhpDocParser\Ast\Type\NullableTypeNode` should not be handled as special case (equals to union `null|type`)

Reply:
?type is parsed as NullableType
type|null is parsed as UnionType
So I handle only UnionType case as list of FullyQualified types because there can be also annotation like that
Product|int|null